### PR TITLE
Core dodge parry

### PR DESCRIPTION
**Pull request by catszeid**
Fix the core code to properly simulate Dodge, Parry, and other segments of the attack table when attacking a player.

### CODE REVIEW

**Review by vigo2 — CHANGES_REQUESTED** [review]

File: sim/core/base_stats_auto_gen.go [hunk 1]
```diff
@@ -7,12 +7,13 @@ package core
...
 const HasteRatingPerHastePercent = 1
 const CritRatingPerCritChance = 1
 const SpellCritRatingPerCritChance = 1
 const MeleeHitRatingPerHitChance = 1
 const SpellHitRatingPerHitChance = 1
-const DefenseRatingPerDefense = 1.500000
+const DefenseRatingPerDefense = 1.0
 const DodgeRatingPerDodgeChance = 13.800001
 const ParryRatingPerParryChance = 13.800001
 const BlockRatingPerBlockChance = 5.000000
```
[the remark this answers is not in the document]
Comment (catszeid) [reply sim/core/base_stats_auto_gen.go:19]: Agreed the overall cleanup should be done in a later PR. Fixed the nit.

File: sim/core/spell_outcome.go [hunk 2]
```diff
@@ -418,7 +419,7 @@ func (result *SpellResult) applyAttackTableDodge(spell *Spell, attackTable *Atta
 		return false
 	}
 
-	*chance += max(0, attackTable.BaseDodgeChance-spell.ExpertisePercentage()-spell.Unit.PseudoStats.DodgeReduction)
+	*chance += max(0, attackTable.BaseDodgeChance-spell.Unit.PseudoStats.DodgeReduction)
```
[the remark this answers is not in the document]
Comment (catszeid) [reply sim/core/spell_outcome.go]: Removed the field and all references of it.

File: sim/warrior/talents.go [hunk 3]
```diff
@@ -15,8 +15,8 @@ func (warrior *Warrior) ToughnessArmorMultiplier() float64 {
 func (warrior *Warrior) ApplyTalents() {
 	warrior.AddStat(stats.MeleeCrit, core.CritRatingPerCritChance*1*float64(warrior.Talents.Cruelty))
 	warrior.ApplyEquipScaling(stats.Armor, warrior.ToughnessArmorMultiplier())
-	warrior.PseudoStats.BaseDodge += 0.01 * float64(warrior.Talents.Anticipation)
-	warrior.PseudoStats.BaseParry += 0.01 * float64(warrior.Talents.Deflection)
+	warrior.AddStat(stats.Dodge, 0.01*float64(warrior.Talents.Anticipation))
```
Comment (TheGroxEmpire) [inline sim/warrior/talents.go:18]: Anticipation should be a defense stat.  Also as what Vigo said, defense stat should be calculated into the attack table.

**Review by TheGroxEmpire — CHANGES_REQUESTED** [review]

File: sim/_paladin/paladin.go [hunk 4]
```diff
@@ -188,8 +188,8 @@ func NewPaladin(character *core.Character, talentsStr string) *Paladin {
 	paladin.AddStatDependency(stats.BonusArmor, stats.Armor, 1)
 
 	// Base dodge is unaffected by Diminishing Returns
-	paladin.PseudoStats.BaseDodge += 0.034943
-	paladin.PseudoStats.BaseParry += 0.05
+	paladin.AddStat(stats.Dodge, 0.034943)
+	paladin.AddStat(stats.Parry, 0.05)
```
[the remark this answers is not in the document]
Comment (vigo2) [reply sim/_paladin/paladin.go:192]: It looks like "base dodge" values for < level 60 are hardly available, so changing these requires some testing. Also: not in this patch.

File: sim/core/base_stats_auto_gen.go [hunk 5]
```diff
@@ -7,12 +7,13 @@ package core
...
 const HasteRatingPerHastePercent = 1
 const CritRatingPerCritChance = 1
 const SpellCritRatingPerCritChance = 1
 const MeleeHitRatingPerHitChance = 1
 const SpellHitRatingPerHitChance = 1
-const DefenseRatingPerDefense = 1.500000
+const DefenseRatingPerDefense = 1
 const DodgeRatingPerDodgeChance = 13.800001
 const ParryRatingPerParryChance = 13.800001
 const BlockRatingPerBlockChance = 5.000000
```
Comment (vigo2) [inline sim/core/base_stats_auto_gen.go:19]: If any of the latter three is still in use, they'd have to be adapted to 1 as well, even here.

File: sim/core/spell_outcome.go [hunk 6]
```diff
@@ -542,9 +543,7 @@ func (result *SpellResult) applyEnemyAttackTableDodge(spell *Spell, attackTable
 	}
 
 	dodgeChance := attackTable.BaseDodgeChance +
-		result.Target.PseudoStats.BaseDodge +
-		result.Target.GetDiminishedDodgeChance() -
-		spell.Unit.PseudoStats.DodgeReduction
+		result.Target.GetStat(stats.Dodge)
```
Comment (vigo2) [inline sim/core/spell_outcome.go:546]: This would need to be multiplied by 1% (0.01) then. Here...

File: sim/core/spell_outcome.go [hunk 7]
```diff
@@ -562,8 +561,7 @@ func (result *SpellResult) applyEnemyAttackTableParry(spell *Spell, attackTable
 	}
 
 	parryChance := attackTable.BaseParryChance +
-		result.Target.PseudoStats.BaseParry +
-		result.Target.GetDiminishedParryChance()
+		result.Target.GetStat(stats.Parry)
```
Comment (vigo2) [inline sim/core/spell_outcome.go:564]: ... and here.

File: sim/core/target.go [hunk 8]
```diff
@@ -329,11 +326,26 @@ func NewAttackTable(attacker *Unit, defender *Unit, weapon *Item) *AttackTable {
...
+		// Apply base Block
+		if defender.PseudoStats.CanBlock {
+			table.BaseBlockChance = 0.05 + levelDelta
+		} else {
+			table.BaseBlockChance = 0
+		}
+
+		table.BaseMissChance = 0.05 + levelDelta
+		table.BaseDodgeChance = 0.05 + levelDelta
+		table.BaseCritChance = 0.05 + levelDelta
```
Comment (vigo2) [inline sim/core/target.go:348]: BaseCritChance must be `0.05 - levelDelta`

File: sim/druid/forms.go [hunk 9]
```diff
@@ -142,7 +142,7 @@ func (druid *Druid) registerCatFormSpell() {
 			druid.AutoAttacks.SetMH(clawWeapon)
 
 			druid.PseudoStats.ThreatMultiplier *= 0.71
-			druid.PseudoStats.BaseDodge += 0.02 * float64(druid.Talents.FelineSwiftness)
+			druid.AddStatDynamic(sim, stats.Dodge, 0.02*float64(druid.Talents.FelineSwiftness))
```
Comment (vigo2) [inline sim/druid/forms.go:145]: `2*float64(druid.Talents.FelineSwiftness)` here (but you'll look for stats.Dodge/stats.Parry usage anyhow ;))

File: sim/warrior/warrior.go [hunk 10]
```diff
@@ -157,14 +157,13 @@ func NewWarrior(character *core.Character, talents string, inputs WarriorInputs)
 	warrior.PseudoStats.CanParry = true
 
 	warrior.AddStatDependency(stats.Agility, stats.MeleeCrit, core.CritPerAgiAtLevel[character.Class][int(warrior.Level)]*core.CritRatingPerCritChance)
-	warrior.AddStatDependency(stats.Agility, stats.Dodge, core.DodgeRatingPerDodgeChance/84.746)
+	warrior.AddStatDependency(stats.Agility, stats.Dodge, core.DodgePerAgiAtLevel[character.Class][int32(warrior.Level)])
```
Comment (vigo2) [inline sim/warrior/warrior.go:160]: Is this "Dodge%PerAgiAtLevel" already?

**Review by vigo2 — CHANGES_REQUESTED** [review]

File: sim/core/target.go (hunk 8 above)
Comment (catszeid) [reply sim/core/target.go:348]: `levelDelta` will be negative if the attacker is higher level than the player target (such as a boss), so addition is correct here.

File: sim/warrior/warrior.go (hunk 10 above)
Comment (catszeid) [reply sim/warrior/warrior.go:160]: It may have been for wrath, but SoD requires per level values for each class.

**Review by TheGroxEmpire — APPROVED** [review]

File: sim/core/target.go (hunk 8 above)
Comment (vigo2) [reply sim/core/target.go:348]: ye, for e.g. a boss attacking player, `levelDelta` will be `-0.006`, (-0.6%) so Base{Miss,Dodge,Parry,Block}Chance will all be `0.044` (4.4%), but BaseCritChance of the boss should be `0.056` (5.6%)

File: sim/rogue/talents.go [hunk 11]
```diff
@@ -15,11 +15,11 @@ func (rogue *Rogue) ApplyTalents() {
...
-	rogue.AddStat(stats.Dodge, core.DodgeRatingPerDodgeChance*1*float64(rogue.Talents.LightningReflexes))
-	rogue.AddStat(stats.Parry, core.ParryRatingPerParryChance*1*float64(rogue.Talents.Deflection))
-	rogue.AddStat(stats.MeleeCrit, core.CritRatingPerCritChance*1*float64(rogue.Talents.Malice))
-	rogue.AddStat(stats.MeleeHit, core.MeleeHitRatingPerHitChance*1*float64(rogue.Talents.Precision))
-	rogue.AddStat(stats.SpellHit, core.SpellHitRatingPerHitChance*1*float64(rogue.Talents.Precision))
+	rogue.AddStat(stats.Dodge, 1*float64(rogue.Talents.LightningReflexes))
+	rogue.AddStat(stats.Parry, 1*float64(rogue.Talents.Deflection))
+	rogue.AddStat(stats.MeleeCrit, 1*float64(rogue.Talents.Malice))
+	rogue.AddStat(stats.MeleeHit, 1*float64(rogue.Talents.Precision))
+	rogue.AddStat(stats.SpellHit, 1*float64(rogue.Talents.Precision))
```
Comment (vigo2) [inline sim/rogue/talents.go:22]: Pre-existing: pre-WotLK Precision doesn't add SpellHit

**Review by vigo2 — APPROVED** [review]
Approve after changes ;)

**Pull request merged**